packages/dataviews/src/components/dataform/index.tsx

```diff
@@ -1,17 +1,37 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
 import type { DataFormProps } from '../../types';
-import { getFormLayout } from '../../dataforms-layouts';
+import { DataFormProvider } from '../dataform-context';
+import { normalizeFields } from '../../normalize-fields';
+import { DataFormLayout } from '../../dataforms-layouts/data-form-layout';
 
 export default function DataForm< Item >( {
 	form,
 	...props
 }: DataFormProps< Item > ) {
-	const layout = getFormLayout( form.type ?? 'regular' );
-	if ( ! layout ) {
+	const normalizedFields = useMemo(
```

Comment:
I'm mostly interested in getting the API right in the review. However, I've also noticed something related to implementation that we could perhaps simplify.

Can we do the following here (pseudo-code):

```js
// Fields are only the visible fields
// and contain all the information they need including the layout.
const fields = normalizeFields( fields, form );
```

By doing the normalization as the first step:

- we don't need the context, because the fields are already passed down as they are needed via props
- we can remove a lot of code that normalizes in the leaf components. Examples: [formregular](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/regular/index.tsx#L53), [formpanel](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/panel/index.tsx#L175), [dataformlayout](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/data-form-layout.tsx#L36), etc.

Reply:
If I understand you correctly, you mean replacing the `form.fields` array with the normalized fields already. So `[ "status" ]` would turn into `[ { ...status field definition } ]`?

I am ok with this approach, but initially avoided it because of the nesting possibility. This means we will have to traverse through the entire field tree initially already, instead of having the react components deal with this naturally.

> we can remove a lot of code that normalizes in the leaf components. Examples: [formregular](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/regular/index.tsx#L53), [formpanel](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/panel/index.tsx#L175), [dataformlayout](https://github.com/WordPress/gutenberg/blob/a707275562bcb7cd82446759e701cc0de978449e/packages/dataviews/src/dataforms-layouts/data-form-layout.tsx#L36), etc.

With this new implementation `FormRegular` and `FormPanel` can actually be removed as they are replaced by `FormRegularField` and `FormPanelField`.
`DataFormLayout` would be the only exception.

Reply:
@oandregal it's important to distinguish `form.fields` from the `fields` prop which represents the "schema" of our data basically. 

The function being called here, normalized the "schema" and doesn't touch `form.fields` at all. We could obviously introduce `normalizeForm` function or something (which is I believe what you're suggesting) but it's important to keep these things distinct. `normalizeFields` is a function that is common between DataViews and DataForms.

Reply:
I agree `normalizeFields` is common across DataViews/DataForm so it shouldn't have the `form` argument and we may need another function or just have some code here. That's fine. My point was about simplifying the implementation. Let me try to rephrase a bit to see if it comes across:

The way I see it, `form.fields` is a filter on `fields` (it's also creating new ones, for example, combined fields). So, if we normalize the fields as early as possible, all internal components can just work with the Fields API (`fields`).

As I mentioned elsewhere, this conversation is not a blocker to for me. I'm mostly interested in getting the API right.

Reply:
> The way I see it, form.fields is a filter on fields (it's also creating new ones, for example, combined fields). So, if we normalize the fields as early as possible, all internal components can just work with the Fields API (fields).

That makes sense, my only thought would be if there could be a case where "a field definition is needed for a field that is not in the rendered list ( `form.fields` )?"
Like:
- A conditional field
     - Should not be needed because of the `isVisible` function, and updating `form.fields` on the fly wouldn't be necessary anymore.
- A field that requires data of another field definition and may need to use `getValue` internally ( the slug comes to mind, as it would be auto generated from the title ). 
     - This would be avoided if we still expose the `fields` definitions anyway.